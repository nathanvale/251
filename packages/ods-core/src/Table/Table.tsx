import React, { useMemo, createContext } from "react";
import clsx from "clsx";
import MuiTable, { TableProps as MuiTableProps } from "@material-ui/core/Table";
import MuiTableContainer from "@material-ui/core/TableContainer";
import { makeStyles } from "@material-ui/core/styles";
import {
  OptionalTrackableProps,
  MuiProps,
  AlignXType,
} from "@origin-digital/ods-types";

const useTableStyles = makeStyles(
  (theme) => ({
    withBorder: {
      borderTop: `1px solid ${theme.palette.grey[200]}`,
      borderCollapse: "collapse",
    },
  }),
  { classNamePrefix: "Table" }
);

const useTableContainerStyles = makeStyles(
  () => ({
    withMaxHeight: ({ maxHeight }: TableContainerStylesProps) => ({
      overflowX: "auto",
      maxHeight: `${
        typeof maxHeight === "number" ? `${maxHeight}px` : maxHeight
      }`,
    }),
    withStickyHeader: {
      overflowX: "initial",
    },
  }),
  {
    classNamePrefix: "TableContainer",
  }
);

export interface TableProps
  extends OptionalTrackableProps,
    MuiProps<MuiTableProps>,
    Pick<MuiTableProps, "children" | "stickyHeader"> {
  "aria-label"?: string;
  alignX?: AlignXType;
  bordered?: boolean;
  hover?: boolean;
  size?: "small" | "medium";
  striped?: boolean;
  maxHeight?: number | string;
}

type TableContextValue = Pick<
  TableProps,
  "alignX" | "bordered" | "hover" | "size" | "striped"
>;

type TableContainerStylesProps = Pick<TableProps, "maxHeight">;

export const TableContext = createContext<TableContextValue>({
  alignX: "left",
  bordered: undefined,
  hover: undefined,
  size: "small",
  striped: undefined,
});

export const Table = ({
  alignX,
  children,
  bordered,
  hover,
  maxHeight,
  size,
  stickyHeader,
  striped,

  muiProps,
  ...props
}: TableProps) => {
  // Prevent re-renders when context values haven't changed
  const tableContextValue = useMemo(
    () => ({ alignX, bordered, hover, size, striped }),
    [alignX, bordered, hover, size, striped]
  );

  const { withBorder } = useTableStyles();
  const { withMaxHeight, withStickyHeader } = useTableContainerStyles({
    maxHeight,
  });

  return (
    <TableContext.Provider value={tableContextValue}>
      <MuiTableContainer
        className={clsx({
          [withStickyHeader]: stickyHeader,
          [withMaxHeight]: maxHeight,
        })}
      >
        <MuiTable
          className={clsx({ [withBorder]: bordered })}
          stickyHeader={stickyHeader}
          {...props}
          {...(muiProps || {})}
        >
          {children}
        </MuiTable>
      </MuiTableContainer>
    </TableContext.Provider>
  );
};

Table.defaultProps = {
  "data-id": "table",
  alignX: "left",
  size: "medium",
};

Table.displayName = "Table";
