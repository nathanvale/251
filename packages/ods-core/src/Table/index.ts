export { Table, defaultProps, TableContext } from "./Table";
export { generateTable } from "./Table.helper";
export type TableProps = import("./Table").TableProps;
