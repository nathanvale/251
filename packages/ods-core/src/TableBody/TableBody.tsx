import React from "react";
import MuiTableBody, {
  TableBodyProps as MuiTableBodyProps,
} from "@material-ui/core/TableBody";
import { OptionalTrackableProps, MuiProps } from "@origin-digital/ods-types";

export interface TableBodyProps
  extends OptionalTrackableProps,
    MuiProps<MuiTableBodyProps>,
    Pick<MuiTableBodyProps, "children"> {}

export const TableBody = ({ children, muiProps, ...props }: TableBodyProps) => {
  return (
    <MuiTableBody {...props} {...(muiProps || {})}>
      {children}
    </MuiTableBody>
  );
};

TableBody.defaultProps = {
  "data-id": "table-body",
};

TableBody.displayName = "TableBody";
