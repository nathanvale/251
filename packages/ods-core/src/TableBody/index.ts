export { TableBody } from "./TableBody";
export type TableBodyProps = import("./TableBody").TableBodyProps;
