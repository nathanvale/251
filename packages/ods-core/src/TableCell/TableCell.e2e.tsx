/* eslint-disable react/display-name */
import React from "react";
import { E2ETests } from "@origin-digital/ods-types";
import { Table, TableBody, TableRow, TableCell, Placeholder } from "..";

export const tests: E2ETests = [
  {
    label: "Default TableCell",
    Code: () => (
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>One</TableCell>
            <TableCell>Two</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    ),
  },
  {
    label: "Horizontal aligned TableCell",
    Code: () => (
      <Table>
        <TableBody>
          <TableRow>
            <TableCell alignX="left">Left</TableCell>
            <TableCell alignX="center">Center</TableCell>
            <TableCell alignX="right">Right</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    ),
  },
  {
    label: "Vertical aligned TableCell",
    Code: () => (
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>
              <Placeholder height="100px" />
            </TableCell>
            <TableCell alignY="top">Top</TableCell>
            <TableCell alignY="center">Center</TableCell>
            <TableCell alignY="bottom">Bottom</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    ),
  },
  {
    label: "Horizontal merged TableCell",
    Code: () => (
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>Single</TableCell>
            <TableCell>Single</TableCell>
          </TableRow>
          <TableRow>
            <TableCell colSpan={2}>Double</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    ),
  },
  {
    label: "Vertical merged TableCell",
    Code: () => (
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>Single</TableCell>
            <TableCell rowSpan={2} alignY="top">
              Double
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Single</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    ),
  },
];
