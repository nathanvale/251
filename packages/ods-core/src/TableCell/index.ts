export { TableCell } from "./TableCell";
export type TableCellProps = import("./TableCell").TableCellProps;
