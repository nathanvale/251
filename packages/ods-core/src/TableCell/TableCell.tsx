import React, { useContext } from "react";
import clsx from "clsx";
import MuiTableCell, {
  TableCellProps as MuiTableCellProps,
} from "@material-ui/core/TableCell";
import { makeStyles } from "@material-ui/core/styles";
import {
  OptionalTrackableProps,
  MuiProps,
  AlignYType,
  AlignXType,
} from "@origin-digital/ods-types";
import { TableProps, TableContext } from "../Table";

interface TableCellStylesProps
  extends Pick<TableCellProps, "alignY">,
    Pick<TableProps, "size" | "textVariant"> {}

const useTableCellStyles = makeStyles(
  (theme) => ({
    withVerticalAlign: ({ alignY }: TableCellStylesProps) => ({
      verticalAlign: alignY,
    }),
    withPadding: ({ size }: TableCellStylesProps) => ({
      padding: `${size === "small" ? "4px 12px" : "12px 16px"}`,
    }),
    withoutFirstLastPadding: {
      "&:first-child": {
        paddingLeft: 0,
      },
      "&:last-child": {
        paddingRight: 0,
      },
    },
    withFontSize: ({ textVariant }: TableCellStylesProps) => {
      return {
        fontSize: `${
          theme.typography[textVariant === "body" ? "body1" : "body2"].fontSize
        }px`,
        lineHeight:
          theme.typography[textVariant === "body" ? "body1" : "body2"]
            .lineHeight,
      };
    },
    withBorder: {
      borderLeft: `1px solid ${theme.palette.grey[200]}`,
      borderRight: `1px solid ${theme.palette.grey[200]}`,
    },
  }),
  { classNamePrefix: "TableCell" }
);

export interface TableCellProps
  extends OptionalTrackableProps,
    MuiProps<MuiTableCellProps>,
    Pick<MuiTableCellProps, "children" | "colSpan" | "rowSpan" | "scope"> {
  alignX?: AlignXType;
  alignY?: AlignYType;
  className?: string;
}

export const TableCell = ({
  children,
  className,
  alignX: alignXProp,
  alignY,
  muiProps,
  ...props
}: TableCellProps) => {
  const { alignX, bordered, hover, size, striped, textVariant } = useContext(
    TableContext
  );
  const {
    withoutFirstLastPadding,
    withPadding,
    withBorder,
    withFontSize,
    withVerticalAlign,
  } = useTableCellStyles({ alignY, textVariant, size });

  return (
    <MuiTableCell
      align={alignXProp || alignX}
      className={clsx(className, withPadding, withVerticalAlign, withFontSize, {
        [withBorder]: bordered,
        [withoutFirstLastPadding]: !bordered && !striped && !hover,
      })}
      {...props}
      {...(muiProps || {})}
    >
      {children}
    </MuiTableCell>
  );
};

TableCell.defaultProps = {
  "data-id": "table-cell",
  alignY: "top",
};

TableCell.displayName = "TableCell";
