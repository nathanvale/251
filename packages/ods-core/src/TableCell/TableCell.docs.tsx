import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Placeholder } from "../Placeholder";
import { Table, generateTable } from "../Table";
import { TableBody } from "../TableBody";
import { TableRow } from "../TableRow";
import { TableCell, TableCellProps } from "./TableCell";

export const docs: ComponentDocs<TableCellProps> = {
  category: "Content",
  componentName: "TableCell",
  description:
    "TableCell is a low-level content component. Open the examples in Playroom to see the behaviour of this component.",
  propDescriptions: {
    alignX: "This is a prop which aligns the cell content horizontally.",
    alignY: "This is a prop which aligns the cell content vertically.",
    className: "Additional class for the table cell.",
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => generateTable(),
    },
    additional: [
      {
        label: "Horizontal align",
        description: `In order to set the alignment in cells pass in "align" with a value.`,
        Code: () => (
          <Table bordered>
            <TableBody>
              <TableRow>
                <TableCell alignX="left">Left</TableCell>
                <TableCell alignX="center">Center</TableCell>
                <TableCell alignX="right">Right</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        ),
      },
      {
        label: "Vertical align",
        description: `In order to set the vertical alignment in cells pass in "valign" with a value.`,
        Code: () => (
          <Table bordered>
            <TableBody>
              <TableRow>
                <TableCell>
                  <Placeholder height="100px" />
                </TableCell>
                <TableCell alignY="top">Top</TableCell>
                <TableCell alignY="center">Center</TableCell>
                <TableCell alignY="bottom">Bottom</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        ),
      },
      {
        label: "Horizontal merging",
        description: `In order merge two or more columns horizontally pass in "colSpan" with a numeric size.`,
        Code: () => (
          <Table bordered>
            <TableBody>
              <TableRow>
                <TableCell>Single</TableCell>
                <TableCell>Single</TableCell>
              </TableRow>
              <TableRow>
                <TableCell colSpan={2}>Double</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        ),
      },
      {
        label: "Vertical merging",
        description: `In order merge two or more columns vertically pass in "rowSpan" with a numeric size.`,
        Code: () => (
          <Table bordered>
            <TableBody>
              <TableRow>
                <TableCell>Single</TableCell>
                <TableCell rowSpan={2} alignY="top">
                  Double
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell>Single</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        ),
      },
    ],
  },
  snippets: [],
};
