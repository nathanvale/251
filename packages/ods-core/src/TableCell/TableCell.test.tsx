import React from "react";
import { render } from "@origin-digital/ods-testing-library";
import { Table, TableProps } from "../Table/Table";
import { TableBody } from "../TableBody/TableBody";
import { TableRow } from "../TableRow/TableRow";
import { TableCell, TableCellProps } from "./TableCell";

const renderTableCell = (
  tableCellProps: Partial<TableCellProps> = {},
  tableProps: Partial<TableProps> = {}
) =>
  render(
    <Table {...tableProps}>
      <TableBody>
        <TableRow>
          <TableCell {...tableCellProps}>Hello world</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  );

test("It renders", () => {
  const { container } = renderTableCell();
  expect(container).toMatchSnapshot();
});

test("It adds horizontal align", () => {
  const { container } = renderTableCell({ alignX: "right" });
  expect(container).toMatchSnapshot();
});

test("It adds vertical align", () => {
  const { container } = renderTableCell({ alignY: "bottom" });
  expect(container).toMatchSnapshot();
});

test("It adds borders", () => {
  const { container } = renderTableCell({}, { bordered: true });
  expect(container).toMatchSnapshot();
});

test("It uses small padding", () => {
  const { container } = renderTableCell({}, { size: "small" });
  expect(container).toMatchSnapshot();
});

test("It uses medium padding", () => {
  const { container } = renderTableCell({}, { size: "medium" });
  expect(container).toMatchSnapshot();
});

test("It does not remove first and last padding", () => {
  const { container } = renderTableCell({}, { striped: true });
  expect(container).toMatchSnapshot();
});

test("It adds class name", () => {
  const className = "myCustomClass";
  const { container, getAllByTestId } = renderTableCell({
    className,
  });
  expect(getAllByTestId("table-cell").pop()?.classList).toContain(className);
  expect(container).toMatchSnapshot();
});
