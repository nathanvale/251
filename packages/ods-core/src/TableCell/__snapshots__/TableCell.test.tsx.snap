// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It adds borders 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root Table-withBorder-100"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-116 TableCell-withPadding-120 TableCell-withVerticalAlign-115 TableCell-withVerticalAlign-119 TableCell-withBorder-118"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It adds horizontal align 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            alignx="right"
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-50 TableCell-withPadding-54 TableCell-withVerticalAlign-49 TableCell-withVerticalAlign-53 TableCell-withoutFirstLastPadding-51 MuiTableCell-alignRight"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It adds vertical align 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-83 TableCell-withPadding-87 TableCell-withVerticalAlign-82 TableCell-withVerticalAlign-86 TableCell-withoutFirstLastPadding-84"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It does not remove first and last padding 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root TableRow-withStripes-207"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-215 TableCell-withPadding-219 TableCell-withVerticalAlign-214 TableCell-withVerticalAlign-218"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It renders 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-17 TableCell-withPadding-21 TableCell-withVerticalAlign-16 TableCell-withVerticalAlign-20 TableCell-withoutFirstLastPadding-18"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It uses medium padding 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-182 TableCell-withPadding-186 TableCell-withVerticalAlign-181 TableCell-withVerticalAlign-185 TableCell-withoutFirstLastPadding-183"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`It uses small padding 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
      data-id="table"
    >
      <tbody
        class="MuiTableBody-root"
        data-id="table-body"
      >
        <tr
          class="MuiTableRow-root"
          data-id="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body TableCell-withPadding-149 TableCell-withPadding-153 TableCell-withVerticalAlign-148 TableCell-withVerticalAlign-152 TableCell-withoutFirstLastPadding-150"
            data-id="table-cell"
          >
            Hello world
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
